neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+
+In essence, this NEP is an extension of [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md), which was focused on splitting one shard into multiple shards. 
+
+We are introducing resharding v2, which supports one shard splitting into two within one epoch at a pre-determined split boundary. The NEP includes performance improvement to make resharding feasible under the current state as well as actual resharding in mainnet and testnet (To be specific, splitting the largest shard into two). 
+
+While the new approach addresses critical limitations left unsolved in NEP-40 and is expected to remain valid for foreseeable future, it does not serve all use cases, such as dynamic resharding. 
+
+## Motivation
+
+Currently, NEAR protocol has four shards. With more partners onboarding, we started seeing that some shards occasionally become over-crowded with respect to total state size and number of transactions. In addition, with state sync and stateless validation, validators will not need to track all shards and validator hardware requirements can be greatly reduced with smaller shard size. With future in-memory tries, it's also important to limit the size of individual shards.
+
+## Specification
+
+### High level assumptions
+
+* Flat storage is enabled.
+* Shard split boundary is predetermined and hardcoded. In other words, necessity of shard splitting is manually decided.
+* For the time being resharding as an event is only going to happen once but we would still like to have the infrastructure in place to handle future resharding events with ease.
+* Merkle Patricia Trie is the underlying data structure for the protocol state.
+* Epoch is at least 6 hrs long for resharding to complete.
+
+### High level requirements
+
+* Resharding must be fast enough so that both state sync and resharding can happen within one epoch.
+* Resharding should work efficiently within the limits of the current hardware requirements for nodes.
+* Potential failures in resharding may require intervention from node operator to recover.
+* No transaction or receipt must be lost during resharding.
+* Resharding must work regardless of number of existing shards.
+* No apps, tools or code should hardcode the number of shards to 4.
+
+### Out of scope
+
+* Dynamic resharding
+  * automatically scheduling resharding based on shard usage/capacity
+  * automatically determining the shard layout
+* Merging shards or boundary adjustments
+* Shard reshuffling
+
+### Required protocol changes
+
+A new protocol version will be introduced specifying the new shard layout which would be picked up by the resharding logic to split the shard.
+
+### Required state changes
+
+* For the duration of the resharding the node will need to maintain a snapshot of the flat state and related columns. As the main database and the snapshot diverge this will cause some extent of storage overhead.
```

Comment:
is there an estimate of how much more storage is required during resharding

Reply:
Added rounded up numbers based on [results shared on zulip](https://near.zulipchat.com/#narrow/stream/407288-pagoda.2Fcore.2Fresharding/topic/hardware.20overhead)